src/Chapter1.hs

```diff
@@ -553,8 +555,15 @@ value after "=" where the condition is true.
 
 Casual reminder about adding top-level type signatures for all functions :)
 -}
+isBetween :: Int -> Int -> Int -> Bool
+isBetween middle upper lower = ((middle > lower) && (middle < upper))
 
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | (isBetween y x z) || (isBetween y z x) = y
+    | (isBetween x y z) || (isBetween x z y) = x
+    | (isBetween z x y) || (isBetween z y x) = z
+    | otherwise = x
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂


It looks like you can move last two cases under a single otherwise and return `z` in here.
The implementation will be a bit shorter 🙂